spec/Candid.md

```diff
@@ -864,142 +864,213 @@ service { <methtype>;* } <: service { <methtype'>;* }
 
 ```
-record { N1':<datatype1'>;* } <: record { N1:<datatype1>;* } ~> f1
-record { N2:<datatype2>;* } <: record { N2':<datatype2'>;* } ~> f2
-------------------------------------------------------------------------------------------------------------------
-func ( <datatype1>,* ) -> ( <datatype2>,* ) <funcann>* <: func ( <datatype1'>,* ) -> ( <datatype2'>,* ) <funcann>*
-  ~> \x.\y.f2 (x (f1 y))
+record {<v>;*} :? record {<t>;*} ~> record {<v'>,*}
+---------------------------------------------------
+(<v>,*) :? (<t>,*) ~> (<v'>,*)
 ```
 
-#### Services
 
-```
+## Properties
 
--------------------------------------------------
-service { <methtype'>;* } <: service { } ~> \x.{}
+The relations above have certain properties. To express them, we need the relation `V : T`, expressing that `V` has inherently type `T`. Instead of defining this relation on its own, we take the first property below as its definition:
 
-<functype> <: <functype'> ~> f1
-service { <methtype>;* } <: service { <methtype'>;* } ~> f2
-------------------------------------------------------------------------------------------------
-service { <name> : <functype>; <methtype>;* } <: service { <name> : <functype'>; <methtype'>;* }
-  ~> \x.{f1 x; <name> = f2 x.<name>}
-```
+* Correctness and completeness of decoding:
+  ```
+  (v : T) ⟺ (∃ v'. v' :? T ~> v)
+  ```
 
+* Roundtripping:
+  ```
+  (v : T) ⟺ v :? T ~> v
+  ```
 
-## Open Questions
+* Uniqueness of decoding:
+  ```
+  v :? T ~> v1, v :? T ~> v2 ⇒ v1 = v2
+  ```
 
-* Support default field values?
-* Support generic type definitions?
-* Namespaces for imports?
+* Soundness of subtyping:
+  ```
+  T <: T' ⇒ ∀ v : T. ∃ v'. v :? T ~> v'
+  ```
+
+* Higher-order soundness of subtyping
+  See <./IDL-Soundness.md>, with the following instanstatiations:
+  ```
+  s1 ~> s2 ≔ s2 <: s1
+  t1 <: t2 ≔ (∀ v. t1: v :? t2 ~> _)
+  s1 in t1 <: s2 in t2 ≔ (to be done)
+  s1 <:h s2 ≔ (host-language dependent)
+  ```
+
+* Transitivity of subtyping:
+  ```
+  T <: T2, T2 <: T3 ⇒ T <: T3
+  ```
+
+* Transitive coherence does not hold:
+  ```
+  T <: T2, T2 <: T3
+  v1 : T
+  v1 :? T3 ~> v3
+  v1 :? T2 ~> v2, v2 :? T3 ~> v3'
+  ```
+  does not imply `v3 = v3'`.
+
+  However, it implies that `R(v3,v3')`, where `R` is the smallest homomorphic, reflexive relation `R` that satisfies `∀ v. R(opt v, null)`.
+
+Other design goals are not satisfied (or hard to express formally?):
+
+* Completeness of subtyping:
+
+  A rigorous formulation of completeness could be that `<:` is the largest transitive and sound relation (in the above sense), i.e. that
+  ```
+  (∀ v. v :? T ~> _ ⇒ v :? T' ~> _) ⇒ T <: T'
+  ```
+  This does not hold as state, because of counter examples involving the empty type. For example we do not have `opt empty <: null`, or `Empty <: t` where `type Empty = rec { Empty }`
```

Comment:
I would expect the following formulation:
```
(∀ v, v'. v : T /\ v :? T' ~> v') ⇒ T <: T'
```
Is there anything wrong with that? And should it not be a requirement?

Reply:
That formulation doesn't make sense to me, the antecent is never true.

Should `/\` be `→`?  Or should `v'` be behind some ∃?

Reply:
Ah, oops, wrong quantifier:
```
(∃ v, v'. v : T /\ v ~> v' : T') ⇒ T <: T'
```

Reply:
Now the antecent is far too weak, it says “`T` is a subtype of `T'` if _some_ values of type `T` make sense at type `T'`). This would, for example, relate all list types (use `v = []` and `v' = []`).

Reply:
How about I remove this section (so this can land if all else is fine), and we discuss this in a separate issue?

Reply:
Eh, you're right, I should turn my brain on. Okay, last try, if that's wrong as well, feel free to remove it. :)
```
(∃ v. v : T /\ ∀ v. v : T ⇒ ∃ v'. v ~> v' : T') ⇒ T <: T'
```
This looks a bit like monster-barring the uninhabited case, but I think it is a natural definition, saying: if serialised values exist for type T, and all such values can be deserialised at type T', then T ought to be considered a subtype of T'.

Uninhabited types tend to make everything harder...

Reply:
Nope, sorry still not good. 

In the antecent instead of `∀ v. v : T ⇒ ` I am sure you  `∀ v. v ~> _ : T ⇒ ` (i.e. not just the “canonical” values, but really “all values that can be deserialized at T). But I think that’s what you mean anyways.

But empty types still break it: This definition would still imply `vec Empty <: vec t` for all `t`.

And even then: Remember that coercion never fails for reference values (because we don't do a subtyping check when decoding and coercion), so any definition for completeness that is based on purely the `_ ~> _ : _` relation will probably relate all function types, which is obviously not sound (in the non-local; IDL-Soundess.md sense).

Let’s take this into a separate issue.

Reply:
Moved to https://github.com/dfinity/candid/issues/132